File: Practical Exam/AOA/Python/1.py
import array

def insertion_sort(arr):
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key

def selection_sort(arr):
    n = len(arr)
    for i in range(n):
        min_index = i
        for j in range(i+1, n):
            if arr[j] < arr[min_index]:
                min_index = j
        arr[i], arr[min_index] = arr[min_index], arr[i]

def dynamic_input():
    n = int(input("Enter the number of elements: "))
    arr = array.array('i', [0] * n)
    print("Enter the elements:")
    for i in range(n):
        arr[i] = int(input())
    return arr

def main():
    arr = dynamic_input()

    print("Select sorting algorithm:")
    print("1. Insertion Sort")
    print("2. Selection Sort")
    choice = int(input("Enter your choice (1 or 2): "))

    if choice == 1:
        insertion_sort(arr)
    elif choice == 2:
        selection_sort(arr)
    else:
        print("Invalid choice")

    if choice in [1, 2]:
        print("Sorted array:")
        for num in arr:
            print(num, end=" ")

if __name__ == "__main__":
    main()

File: Practical Exam/AOA/C/1.c
#include <stdio.h>
#include <stdlib.h>

// Function to perform insertion sort
void insertionSort(int arr[], int n) {
    int i, key, j;
    for (i = 1; i < n; i++) {
        key = arr[i];
        j = i - 1;
        
        /* Move elements of arr[0..i-1], that are
           greater than key, to one position ahead
           of their current position */
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j = j - 1;
        }
        arr[j + 1] = key;
    }
}

// Function to perform selection sort
void selectionSort(int arr[], int n) {
    int i, j, min_idx;
    // One by one move boundary of unsorted subarray
    for (i = 0; i < n-1; i++) {
        // Find the minimum element in unsorted array
        min_idx = i;
        for (j = i+1; j < n; j++)
            if (arr[j] < arr[min_idx])
                min_idx = j;
        // Swap the found minimum element with the first element
        int temp = arr[min_idx];
        arr[min_idx] = arr[i];
        arr[i] = temp;
    }
}

int main() {
    int n, choice;
    printf("Enter the number of elements: ");
    scanf("%d", &n);
    
    int *arr = (int *)malloc(n * sizeof(int));
    
    printf("Enter %d integers:\n", n);
    for (int i = 0; i < n; i++)
        scanf("%d", &arr[i]);
    
    printf("Choose sorting algorithm:\n");
    printf("1. Insertion Sort\n");
    printf("2. Selection Sort\n");
    printf("Enter your choice: ");
    scanf("%d", &choice);
    
    switch (choice) {
        case 1:
            insertionSort(arr, n);
            printf("Array sorted using Insertion Sort: \n");
            break;
        case 2:
            selectionSort(arr, n);
            printf("Array sorted using Selection Sort: \n");
            break;
        default:
            printf("Invalid choice\n");
            return 1;
    }
    
    // Print the sorted array
    for (int i = 0; i < n; i++)
        printf("%d ", arr[i]);
    printf("\n");

    // Free dynamically allocated memory
    free(arr);

    return 0;
}